riscv/src/code_gen.rs

```diff
@@ -539,16 +537,18 @@ fn memory(with_bootloader: bool) -> String {
 pub trait InstructionArgs {
     type Error: fmt::Display;
 
-    fn l(&self) -> Result<String, Self::Error>;
+    fn l(&self) -> Result<impl AsRef<str>, Self::Error>;
     fn r(&self) -> Result<Register, Self::Error>;
     fn rri(&self) -> Result<(Register, Register, u32), Self::Error>;
+    /// Returns the usual rd, rs1, rs2
     fn rrr(&self) -> Result<(Register, Register, Register), Self::Error>;
+    /// Special case used in amo* instructions, returning rd, rs2, rs1
+    fn rrr2(&self) -> Result<(Register, Register, Register), Self::Error>;
     fn ri(&self) -> Result<(Register, u32), Self::Error>;
     fn rr(&self) -> Result<(Register, Register), Self::Error>;
-    fn rrl(&self) -> Result<(Register, Register, String), Self::Error>;
-    fn rl(&self) -> Result<(Register, String), Self::Error>;
+    fn rrl(&self) -> Result<(Register, Register, impl AsRef<str>), Self::Error>;
+    fn rl(&self) -> Result<(Register, impl AsRef<str>), Self::Error>;
     fn rro(&self) -> Result<(Register, Register, u32), Self::Error>;
-    fn rrro(&self) -> Result<(Register, Register, Register, u32), Self::Error>;
```

Comment:
`rrro` stands for "register, register, register, offset", but there is no RISC-V instruction that encodes this much information (there is not enough space in the 32 bits). It existed because the assembly syntax for these instructions allowed for writing `amoadd.d rd, rs2, 0(rs1)`, but there is no place to put that `0` in the actual encoding (if I try to assemble with any other value I get an error).

So, at first I replaced `rrro()` with `rrr()`, and changed it to accept both `None` and 0 offset. Then I figured out that the order of the registers expected by these instructions was different from other instructions using `rrr()`, so I created `rrr2()` (ok, the name could be better) to say that this is an instruction that expects 3 registers, but in a different order from `rrr()`.

Reply:
i see, i think name is fine, can you add a small comment?

Reply:
There are comments on both `rrr()` and `rrr2()` explaining the difference between them.